src/net.h

```diff
@@ -417,6 +419,187 @@ class V1Transport final : public Transport
+     *     v                                              v         |
+     *    KEY -> GARB_GARBTERM -> GARBAUTH -> VERSION -> APP -> APP_READY
+     */
+    enum class RecvState : uint8_t {
+        /** Public key.
+         *
+         * This is the initial state, during which the other side's public key is
+         * received. When that information arrives, the ciphers get initialized and the state
+         * becomes GARB_GARBTERM. */
+        KEY,
+
+        /** Garbage and garbage terminator.
+         *
+         * Whenever a byte is received, the last 16 bytes are compared with the expected garbage
+         * terminator. When that happens, the state becomes GARBAUTH. If no matching terminator is
+         * received in 4111 bytes (4095 for the maximum garbage length, and 16 bytes for the
+         * terminator), the connection aborts. */
+        GARB_GARBTERM,
+
+        /** Garbage authentication packet.
+         *
+         * A packet is received, and decrypted/verified with AAD set to the garbage received during
+         * the GARB_GARBTERM state. If that succeeds, the state becomes VERSION. If it fails the
+         * connection aborts. */
+        GARBAUTH,
+
+        /** Version packet.
+         *
+         * A packet is received, and decrypted/verified. If that succeeds, the state becomes APP,
+         * and the decrypted contents is interpreted as version negotiation (currently, that means
+         * ignoring it, but it can be used for negotiating future extensions). If it fails, the
+         * connection aborts. */
+        VERSION,
+
+        /** Application packet.
+         *
+         * A packet is received, and decrypted/verified. If that succeeds, the state becomes
+         * APP_READY and the decrypted contents is kept in m_recv_decode_buffer until it is
+         * retrieved as a message by GetMessage(). */
+        APP,
+
+        /** Nothing (an application packet is available for GetMessage()).
+         *
+         * Nothing can be received in this state. When the message is retrieved by GetMessage,
+         * the state becomes APP again. */
+        APP_READY,
+    };
+
+    /** State type that defines the contents of the send buffer.
+     *
+     * Diagram:
+     *
+     *   start                                       /----------\
+     *     |                                         |          |
+     *     v                                         v          |
+     *    KEY -> KEY_GARBTERM_GARBAUTH_VERSION -> APP_READY -> APP
+     */
+    enum class SendState : uint8_t {
+        /** Public key.
+         *
+         * This is the initial state. The public key is sent out. When the receiver
+         * receives the other side's public key and transitions to GARB_GARBTERM, the sender state
+         * becomes KEY_GARBTERM_GARBAUTH_VERSION. The key is left in the send buffer when this
+         * happens, because it may not have been fully sent out yet. */
```

Comment:
fb3ec8fd9e96f6c2d16db02fd2d683ea097a90ff: why not remain in the `KEY` state until we've completely sent out the public key? And only then transition to `GARBTERM_GARBAUTH_VERSION`?

Reply:
Perhaps a more natural border would be `KEY_GARB` -> `GARBTERM_GARBAUTH_VERSION` then? But that sounds like what you implemented in the branch. Maybe just leave it as is, I might revisit this after I've looked at more of the code.

Reply:
I suppose the big conceptual difference between our send and receive state machines is that the latter has a clearly defined moment when the key exchange is complete, i.e. once it received the public key bytes. From the `send` state machine point of view this happens in the background at some unpredictable point.